Reimplement as a blueprint

mshima opened the pull request:
~Depends on https://github.com/jhipster/generator-jhipster/pull/17629.~

Closes https://github.com/jhipster/generator-jhipster-ionic/issues/604.

mshima commented:
As a blueprint:
```
jhipster-ionic jdl bug-tracker.jh
```
or
```
jhipsterjdl bug-tracker.jh --blueprints ionic
```

Standalone (no entity support):
```
jhipster-ionic --base-name ionic4j --authentication-type jwt --standalone
```
or
```
jhipster ionic --blueprints ionic --base-name ionic4j --authentication-type jwt --standalone
```

mraible commented:
This is awesome, @mshima! I plan to do some testing in the next week and provide feedback.

mraible commented:
I tried this today. The following command doesn't work.

```
$ npm install -g mshima/generator-jhipster-ionic#esm
zsh: no matches found: mshima/generator-jhipster-ionic#esm
```

I was able to check out your branch and run `npm link` on it.

This doesn't work either:

```
$ jhipster-ionic jdl bug-tracker.jh --skip-jhipster-dependencies
zsh: command not found: jhipster-ionic
```

I originally had the CLI named `ionic4j`. Using that doesn't work either.

This command does work:

```
jhipster jdl bug-tracker.jh --blueprints ionic --skip-jhipster-dependencies
```

However, it prompts the user to generate the app in `../ionic4j` by default. Previously, it was `ionic4j` because the expectation is you'd create it next to your JHipster app, you wouldn't run the command from inside your app. 

After continuing, I realize it's generating the JHipster app and the Ionic app at the same time. That's pretty slick. This was never possible before. We've mostly expected people to have an app and they want to add an Ionic client to it. This is how the [React Native blueprint](https://github.com/jhipster/generator-jhipster-react-native) currently works too. It expects your JHipster app to already exist. 

To make sure I'm correct about ^^, I'll tag @ruddell in this comment.

mshima commented:
> I tried this today. The following command doesn't work.
> 
> ```
> $ npm install -g mshima/generator-jhipster-ionic#esm
> zsh: no matches found: mshima/generator-jhipster-ionic#esm
> ```

I've removed the generator-jhipster branch it was depending on.
Updated and should be working now.

> After continuing, I realize it's generating the JHipster app and the Ionic app at the same time. That's pretty slick. This was never possible before. We've mostly expected people to have an app and they want to add an Ionic client to it.

Moving https://github.com/mshima/generator-jhipster-ionic/blob/51ed2727eeaff14169a45b7734efd2e09830ab59/generators/app/generator.mjs#L56-L68 to install priority will make the ionic app to be generated after the base application, so it's possible to show a different logo and so one.

`jhipster ionic` or `jhipster-ionic` expects a existing application `jhipster --blueprints ionic` no (it generates the base application and ionic application).

mraible commented:
@mshima Can you please rebase so I can try again?

mraible commented:
I tried installing this today with `npm install -g "mshima/generator-jhipster-ionic#esm"` and it errors out with:

```
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uuid@3.3.2: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN cleanup Failed to remove some directories [
npm WARN cleanup   [
npm WARN cleanup     '/usr/local/lib/node_modules/generator-jhipster-ionic/node_modules',
npm WARN cleanup     [Error: EINVAL: invalid argument, rmdir '/usr/local/lib/node_modules/generator-jhipster-ionic/node_modules/generator-jhipster/generators/server/templates'] {
npm WARN cleanup       errno: -22,
npm WARN cleanup       code: 'EINVAL',
npm WARN cleanup       syscall: 'rmdir',
npm WARN cleanup       path: '/usr/local/lib/node_modules/generator-jhipster-ionic/node_modules/generator-jhipster/generators/server/templates'
npm WARN cleanup     }
npm WARN cleanup   ]
npm WARN cleanup ]
npm ERR! code 236
npm ERR! git dep preparation failed
npm ERR! command /usr/local/bin/node /usr/local/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/Users/mraible/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit --include=dev --include=peer --include=optional --no-package-lock-only --no-dry-run
npm ERR! npm WARN using --force Recommended protections disabled.
npm ERR! npm ERR! code ENOTDIR
npm ERR! npm ERR! syscall rename
npm ERR! npm ERR! path /usr/local/lib/node_modules/generator-jhipster-ionic
npm ERR! npm ERR! dest /usr/local/lib/node_modules/.generator-jhipster-ionic-nJwm2eI1
npm ERR! npm ERR! errno -20
npm ERR! npm ERR! ENOTDIR: not a directory, rename '/usr/local/lib/node_modules/generator-jhipster-ionic' -> '/usr/local/lib/node_modules/.generator-jhipster-ionic-nJwm2eI1'
npm ERR!
npm ERR! npm ERR! A complete log of this run can be found in:
npm ERR! npm ERR!     /Users/mraible/.npm/_logs/2022-03-24T17_33_27_965Z-debug-0.log

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/mraible/.npm/_logs/2022-03-24T17_33_06_346Z-debug-0.log
```

I ran removed `/usr/local/lib/node_modules/generator-jhipster-ionic/node_modules` and tried again. It still happens.

mshima commented:
That's weird, I can reproduce, don't know how to fix.

mraible commented:
I tried it with clone, `npm i && npm link` and got a bit further. 

`jhipster ionic` doesn't work, but `jhipster-ionic` does.

```
$ jhipster ionic
INFO! Using bundled JHipster
ionic is not a known command. See 'jhipster --help'.
```

Then it prompts me for the backend. 

```
? Where JHipster backend application is located? backend
⬢ Welcome to the JHipster Project Name ⬢
? What is the project name of your application? Flickr2
? What is the default indentation? 2
```

This is good, but the default (`../backend`) should be `backend` since that's what previous versions used.

Why is "⬢ Welcome to the JHipster Project Name ⬢" printed? I believe this should be suppressed.

"What is the project name" looks kinda funny. JHipster uses "What is the base name of your application?". Maybe we should just change it to "What is the name"?

Why prompt for indentation? I'd just default to 2 spaces.

It also doesn't create the project directory, it just generates it in the current directory. While I'd like to change so it's like before (and creates the directory), I noticed that the [React Native blueprint](https://dev.to/ruddell/jhipster-react-native-demo-1c54) requires you to create a directory and cd into it first, so I guess this is the way.

mraible commented on .eslintignore:
```diff
@@ -1,4 +1,2 @@
-node_modules
-travis
+**/templates/
+**/resources/
```
Shouldn't this still include `node_modules`? Or is it excluded by default?

mshima replied:
By default

mraible commented on .github/workflows/generator.yml:
```diff
@@ -0,0 +1,14 @@
+jobs:
+  ionic:
+    name: npm-test
+    runs-on: ubuntu-20.04
```
```suggestion
    runs-on: ubuntu-latest
```

mraible commented on .github/workflows/generator.yml:
```diff
@@ -0,0 +1,14 @@
+    runs-on: ubuntu-20.04
+    timeout-minutes: 10
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v2
+        with:
```
```suggestion
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
```

mraible commented on .github/workflows/ionic.yml:
```diff
@@ -20,67 +20,62 @@
       - uses: actions/checkout@v3
+        with:
+          path: generator-jhipster-ionic
+      - name: Link ionic blueprint
```
```suggestion
      - name: Link Ionic blueprint
```

mraible commented on .github/workflows/publish.yml:
```diff
@@ -11,9 +11,9 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-node@v2.5.0
```
```suggestion
      - uses: actions/setup-node@v3
```

mraible commented on .yo-rc.json:
```diff
@@ -1,10 +1,34 @@
+        "written": true
+      }
+    },
+    "jhipsterVersion": "7.5.0",
```
Should this be updated to 7.7.0?

mshima replied:
Can be removed

mshima replied:
Oops, this is the generated. Let’s keep 7.7.0

mraible commented on .yo-rc.json:
```diff
@@ -1,10 +1,34 @@
+    },
+    "jhipsterVersion": "7.5.0",
+    "prettierDefaultIndent": 2,
+    "projectName": "JHipster Ionic Blueprint v2",
```
Do we need to include the version number? Technically, it's 1.0 since the previous version was a module, not a blueprint. I like calling it 2.0 though!

mraible commented on generators/app/generator.mjs:
```diff
@@ -0,0 +1,69 @@
+    super(args, opts, { taskPrefix: PRIORITY_PREFIX, ...features });
+
+    this.option('ionic-dir', {
+      desc: 'Directory for jhipster application',
```
```suggestion
      desc: 'Directory of JHipster application',
```

mraible commented on generators/app/generator.mjs:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    if (this.blueprintConfig.appDir) {
+      throw new Error('jhipster-ionic:app generator must run at backend application folder');
```
```suggestion
      throw new Error('jhipster-ionic:app generator must run in backend application directory');
```

mraible commented on generators/app/generator.mjs:
```diff
@@ -0,0 +1,69 @@
+            {
+              type: 'input',
+              name: 'ionicDir',
+              message: 'Where do you want to generate ionic application?',
```
```suggestion
              message: 'Where do you want to generate an Ionic application?',
```
Does this allow creating a directory and then the app in it instead of requiring you to create the directory first?

mshima replied:
Yes we generate the directory.

mraible commented on generators/ionic/resources/base/LICENSE:
```diff
@@ -0,0 +1,201 @@
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright 2018-Present Okta, Inc.
```
We might want to change this and put JHipster's copyright info at the top like other repos have.

mshima replied:
Not sure if makes sense to generate LICENSE at the generated application.
Unless we want to force the generated code to have the license.
Is JHipster generated code copyrighted?

mraible replied:
I agree. Let's remove the generated license.

mraible commented on generators/ionic/resources/base/README.md:
```diff
@@ -0,0 +1,159 @@
+# The Ionic JHipster Starter 🤓
```
Will this file be included as the README in the new project? If so, we should adjust it and add JHipster commands to it.

mshima replied:
Yes it will

mraible commented on generators/ionic/resources/base/package.json:
```diff
@@ -0,0 +1,147 @@
+{
+  "version": "0.0.1",
+  "private": true,
+  "description": "A hipster Ionic project, made with 💙 by @oktadev!",
```
```suggestion
  "description": "A hipster Ionic project, made with 💙 by @jhipster!",
```

mraible commented on generators/ionic/generator.mjs:
```diff
@@ -0,0 +1,458 @@
+    });
+
+    this.option('app-dir', {
+      desc: 'Directory for jhipster application',
```
```suggestion
      desc: 'Directory for JHipster application',
```

mraible commented on generators/ionic/generator.mjs:
```diff
@@ -0,0 +1,458 @@
+    if (this.options.help) return;
+
+    if (this.blueprintConfig.ionicDir) {
+      throw new Error('Ionic generator must run at ionic application folder');
```
```suggestion
      throw new Error('Ionic generator must run in Ionic application directory');
```

mraible commented on generators/ionic/generator.mjs:
```diff
@@ -0,0 +1,458 @@
+        {
+          type: 'input',
+          name: 'appDir',
+          message: 'Where JHipster backend application is located?',
```
```suggestion
          message: 'Enter the directory where your JHipster app is located:',
```
To match previous version so docs don't need updating.

mraible replied:
Where is the `What is the project name of your application?` prompt defined? I'd like to change it to match the previous: `What do you want to name your Ionic application?`

mraible replied:
And how do we suppress the "⬢ Welcome to the JHipster Project Name ⬢" message?

mshima replied:
I am trying to reuse basic generators, I will change at generator-jhipster side.

mraible replied:
Does that mean this will require a new release of JHipster for these changes?

mshima replied:
Should be fixed.

mshima commented:
I don’t have write permission. So do it at will.

mraible commented:
@mshima I added you as a maintainer. It looks like @jdubois needs to add you as a team member. I don't have permission to do this.

jdubois commented:
I invited @mshima to https://github.com/orgs/jhipster/teams/jhipster-ionic-developers - is that good for you?

mraible commented:
Yes, thank you.
